### PR TITLE
feat: support merging the "skip" attribute from included files.

### LINKED ISSUE #3224: Merge the "skip" attribute from included terragrunt files is not supported
## Describe the bug

The current Terragrunt documentation, specifically the section on the skip attribute [here](https://terragrunt.gruntwork.io/docs/reference/config-blocks-and-attributes/#skip), stipulates that this flag must be explicitly set within Terragrunt modules that are intended to be skipped. This implies that the configuration will not be inherited from include blocks. Despite this being clearly documented, it appears to contradict the merging concept supported by the include block.

From a user's standpoint, it would be beneficial to have the ability to inherit this flag/attribute, similar to how many other attributes are inherited. Is there a specific rationale behind the distinct treatment of this attribute that I may be overlooking?

## Steps To Reproduce

Create a partial terragrunt hcl file

partial.hcl
```hcl
locals {
  module_path_parts = split("/", path_relative_to_include())
  module_base_path  = local.module_path_parts[0]
  module_name       = join("_", slice(local.module_path_parts, 1, length(local.module_path_parts)))
}

skip = !endswith(local.module_name, "_account_layer")
```

Include the partial file in a live module

partial.hcl
```hcl
include {
  path   = find_in_parent_folders("partial.hcl")
  expose = true
}
```

## Expected behavior

As you can see, the `skip` in the partial terragrunt file does not contain a simple `true` or `false`, it contains a more advanced logic that I would like to inherit in included modules, without having to duplicate the logic there, currently that is not possible because the `skip` attribute is not included.

## Nice to haves

- [ ] Terminal output
- [ ] Screenshots

## Versions

- Terragrunt version: [v0.59.5](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.59.5)
- OpenTofu/Terraform version: 1.5.5
- Environment details (Ubuntu 20.04, Windows 10, etc.): macOS 14.x

### PR DESCRIPTION
**Pull request by rodrigorfk**
## Description

Fixes #3224.

This PR is introducing support for merging the content of the `skip` attribute from `include` blocks, in the same way of many other attributes and blocks are already being merged.

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [x] Update the docs.
- [x] Run the relevant tests successfully, including pre-commit checks.
- [x] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [x] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

Added / Removed / Updated [X].

### CODE REVIEW

**Comment by denis256** [comment]
```
-- FAIL: TestResolveTerraformModulesReadConfigFromParentConfig (0.02s)

Expected :config.TerragruntConfig{Catalog:(*config.CatalogConfig)(nil), Terraform:(*config.TerraformConfig)(0xc0006abf10), TerraformBinary:"", TerraformVersionConstraint:"", TerragruntVersionConstraint:"", RemoteState:(*remote.RemoteState)(nil), Dependencies:( ...

Actual   :config.TerragruntConfig{Catalog:(*config.CatalogConfig)(nil), Terraform:(*config.TerraformConfig)(0xc000129260), TerraformBinary:"", TerraformVersionConstraint:"", TerragruntVersionConstraint:"", RemoteState:(*remote.RemoteState)(nil), Dependencies:( ...

```

**Comment by denis256** [comment]
Integration test `TestApplyAllSkipTrue` also seems to fail 

```
    integration_test.go:4314: Copying fixture-skip/ to /tmp/terragrunt-test2863889219
    integration_test.go:4268: [terragrunt apply-all --terragrunt-non-interactive --terragrunt-working-dir /tmp/terragrunt-test2863889219/fixture-skip/skip-true --terragrunt-log-level info]
    integration_test.go:2010: [TestApplyAllSkipTrue] Full contents of show stdout:
    integration_test.go:2010: [TestApplyAllSkipTrue] 
    integration_test.go:2011: [TestApplyAllSkipTrue] Full contents of show stderr:
    integration_test.go:2011: [TestApplyAllSkipTrue] time=2024-06-24T15:42:21+01:00 level=warning msg='apply-all' is deprecated. Running 'terragrunt run-all apply' instead. Please update your workflows to use 'terragrunt run-all apply', as 'apply-all' may be removed in the future!
    integration_test.go:2011: [TestApplyAllSkipTrue] 
    integration_test.go:2011: [TestApplyAllSkipTrue] time=2024-06-24T15:42:21+01:00 level=info msg=The stack at /tmp/terragrunt-test2863889219/fixture-skip/skip-true will be processed in the following order for command apply:
    integration_test.go:2011: [TestApplyAllSkipTrue] Group 1
    integration_test.go:2011: [TestApplyAllSkipTrue] - Module /tmp/terragrunt-test2863889219/fixture-skip/skip-true
    integration_test.go:2011: [TestApplyAllSkipTrue] - Module /tmp/terragrunt-test2863889219/fixture-skip/skip-true/resource1
    integration_test.go:2011: [TestApplyAllSkipTrue] - Module /tmp/terragrunt-test2863889219/fixture-skip/skip-true/resource2
    integration_test.go:2011: [TestApplyAllSkipTrue] 
    integration_test.go:2011: [TestApplyAllSkipTrue] 
    integration_test.go:2011: [TestApplyAllSkipTrue] time=2024-06-24T15:42:21+01:00 level=info msg=Skipping terragrunt module /tmp/terragrunt-test2863889219/fixture-skip/skip-true/terragrunt.hcl due to skip = true. prefix=[/tmp/terragrunt-test2863889219/fixture-skip/skip-true] 
    integration_test.go:2011: [TestApplyAllSkipTrue] time=2024-06-24T15:42:21+01:00 level=info msg=Skipping terragrunt module /tmp/terragrunt-test2863889219/fixture-skip/skip-true/resource2/terragrunt.hcl due to skip = true. prefix=[/tmp/terragrunt-test2863889219/fixture-skip/skip-true/resource2] 
    integration_test.go:2011: [TestApplyAllSkipTrue] time=2024-06-24T15:42:21+01:00 level=info msg=Skipping terragrunt module /tmp/terragrunt-test2863889219/fixture-skip/skip-true/resource1/terragrunt.hcl due to skip = true. prefix=[/tmp/terragrunt-test2863889219/fixture-skip/skip-true/resource1] 
    integration_test.go:2011: [TestApplyAllSkipTrue] 
    integration_test.go:2018: 
        	Error Trace:	/home/denis256/projects/gruntwork/terragrunt/test/integration_test.go:2018
        	Error:      	"" does not contain "hello, Ernie"
        	Test:       	TestApplyAllSkipTrue
    integration_test.go:2019: 
        	Error Trace:	/home/denis256/projects/gruntwork/terragrunt/test/integration_test.go:2019
        	Error:      	"" does not contain "hello, Bert"
        	Test:       	TestApplyAllSkipTrue

```

File: docs/_docs/04_reference/config-blocks-and-attributes.md [hunk 1]
```diff
@@ -1285,9 +1285,8 @@ root level `terragrunt.hcl` file, you can set `skip = true`:
 skip = true
 ```
 
-The `skip` flag must be set explicitly in terragrunt modules that should be skipped. If you set `skip = true` in a
-`terragrunt.hcl` file that is included by another `terragrunt.hcl` file, only the `terragrunt.hcl` file that explicitly
-set `skip = true` will be skipped.
+The `skip` flag can be inherited from an included `terragrunt.hcl` file if `skip` is defined there, unless it is
```
Comment (denis256) [inline docs/_docs/04_reference/config-blocks-and-attributes.md:1288]: I think may be helpful to add an integration test for this case, when the skip flag is inherited from include

**Comment by rodrigorfk** [comment]
Thanks for publishing the failed tests here @denis256 and sorry for not catching them initially, I've fixed them.

**Comment by rodrigorfk** [comment]
Hi @denis256 , could you please share the reason why the `test_signing` job failed? looking into it's definition it seems to only be related to binary signing and not really related to the specific change I am doing in this PR, right? thanks in advance.

**Comment by denis256** [comment]
Hi,
it is an internal issue with Apple... I hope will be fixed soon

🍎 🪱

**Comment by yhakbar** [comment]
Hey @rodrigorfk ,

Do you mind rebasing this PR? I'd like to avoid having it go stale.

**Comment by rodrigorfk** [comment]
done @yhakbar, the latest changes from master were merged into this branch.

**Comment by yhakbar** [comment]
Hey @rodrigorfk ,

The PR is failing CI checks. Did you run checks locally?

You should be able to run `make install-lint && make run-lint` without needing access to AWS or anything.

We're working on making it easier to run tests locally. I know it's not as straightforward as it could be. If you could make sure to execute the tests that you're modifying though, that will speed up the process on this review.

**Comment by rodrigorfk** [comment]
Hey @yhakbar,

Sorry for the issue, forgot to check it locally after the merge, mostly because as you said, it is pretty hard to run all the test suite locally, but anyway, it shouldn't be an excuse for not running focused ones, I've fixed it and I hope it works now.

**Comment by denis256** [comment]
Looks like some tests are still failing:
TestResolveTerraformModulesReadConfigFromParentConfig
TestTableTagging

**Comment by yhakbar** [comment]
Sorry we haven't been able to get this PR merged, @rodrigorfk .

To avoid having it stay stale, would you mind resolving conflicts, then ensure that the tests @denis256 mentioned pass locally?

**Comment by yhakbar** [comment]
This pull request is still stale. I'm going to close it out.

If you would like to open it again, please make sure to rebase, lint and test locally, then request re-review.

**Comment by asvinours** [comment]
this feature is introducing a major breaking change for all existing setup, if the parent file has a `skip` flag, every single module is now being skipped. This means terragrunt has become a no-op for a lot of our repositories over night

**Comment by yhakbar** [comment]
@asvinours ,

It is a breaking change, and was called out in [the release notes](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.68.0). 

Why does the included configuration define a `skip`? Can you give some context as to why you have it defined there, but don't want it included?

**Comment by asvinours** [comment]
We use a parent/child structure for our terragrunt config so we can centralize a lot of the shared logic in the parent file and then include it in all the child files. Each child file correspond to an actual infrastructure module, whereas the parent file is just there to be included and isn't tied to any infra code.

Once I saw terragrunt was skipping all the work I looked at the most recent releases and noticed the changelog but the issue here is terragrunt version doesn't use the major flag of semver, only minor and patch. So as soon as 0.68.0 was released our CICD pipelines pulled down the latest version and our pipelines broke :( 

We're updating our version locks to `< 0.68.0` for now until we have time to adapt all our code to this new behaviour

**Pull request merged**